fedora_messaging/tests/integration/test_twisted_api.py

```diff
@@ -0,0 +1,579 @@
+    explicitly cancels the consumer (by deleting the queue).
+    """
+    queue = str(uuid.uuid4())
+    queues = {queue: {"auto_delete": False, "arguments": {"x-expires": 60 * 1000}}}
+    bindings = [{"queue": queue, "exchange": "amq.topic", "routing_keys": ["#"]}]
+
+    consumers = yield api.twisted_consume(lambda x: x, bindings, queues)
+
+    # Delete the queue and assert the consumer errbacks
+    url = "{base}queues/%2F/{queue}".format(base=HTTP_API, queue=queue)
+    yield treq.delete(url, auth=HTTP_AUTH, timeout=3)
+
+    consumers[0].result.addTimeout(10, reactor)
+    try:
+        yield consumers[0].result
+        pytest.fail("Consumer did not errback!")
+    except exceptions.ConsumerCanceled:
+        pass
+    except defer.TimeoutError:
+        pytest.fail("Timeout reached without consumer calling its errback!")
+
+
+@pytest_twisted.inlineCallbacks
+def test_twisted_consume_permission_denied(admin_user):
+    """
+    Assert the call to api.twisted_consume errbacks on permissions errors
+    """
+    url = "{base}permissions/%2F/{user}".format(base=HTTP_API, user=admin_user)
+    body = {"configure": ".*", "write": "", "read": ""}
+    resp = yield treq.put(url, json=body, auth=HTTP_AUTH, timeout=3)
+    assert resp.code == 204
+
+    queue = str(uuid.uuid4())
+    queues = {queue: {"auto_delete": False, "arguments": {"x-expires": 60 * 1000}}}
+    bindings = [{"queue": queue, "exchange": "amq.topic", "routing_keys": ["#"]}]
+
+    amqp_url = "amqp://{user}:guest@localhost:5672/%2F".format(user=admin_user)
+    try:
+        with mock.patch.dict(config.conf, {"amqp_url": amqp_url}):
+            yield api.twisted_consume(lambda x: x, bindings, queues)
+        pytest.fail("Call failed to raise an exception")
+    except exceptions.BadDeclaration as e:
+        assert e.reason.args[0] == 403
+        assert e.reason.args[1] == (
+            "ACCESS_REFUSED - access to queue '{}' in vhost '/' refused for user"
+            " '{}'".format(queue, admin_user)
+        )
+
+
+@pytest_twisted.inlineCallbacks
+def test_twisted_consume_invalid_message():
+    """Assert messages that fail validation are nacked."""
+    queue = str(uuid.uuid4())
+    queues = {queue: {"auto_delete": False, "arguments": {"x-expires": 60 * 1000}}}
+    bindings = [
+        {
+            "queue": queue,
+            "exchange": "amq.topic",
+            "routing_keys": ["test_twisted_invalid_message"],
+        }
+    ]
+
+    consumers = yield api.twisted_consume(
+        lambda x: pytest.fail("Message should be nacked"), bindings, queues
+    )
+
+    url = HTTP_API + "exchanges/%2F/amq.topic/publish"
+    body = {
+        "properties": {},
+        "routing_key": "test_twisted_invalid_message",
+        "payload": "not json",
+        "payload_encoding": "string",
+    }
+    response = yield treq.post(url, json=body, auth=HTTP_AUTH, timeout=3)
+    response = yield response.json()
+    assert response["routed"] is True
+    consumers[0].cancel()
```

Comment:
`Consumer.cancel()` seems to return a `Deferred`, we should yield it here. Otherwise I don't think we have any guarantee that the consumer will actually process the incoming invalid message, and this test will always succeed.